bin/oneoff/migrate_old_workshops.rb

```diff
@@ -0,0 +1,166 @@
+  },
+  Pd::SharedWorkshopConstants::COURSE_CSF => {
+    nil => Pd::SharedWorkshopConstants::SUBJECT_CSF_101
+  }
+}.freeze
+
+def get_or_create_enrollment(pd_workshop, teacher, created_at=nil)
+  if Pd::Enrollment.exists?(pd_workshop_id: pd_workshop.id, user_id: teacher.id)
+    return Pd::Enrollment.find_by(pd_workshop_id: pd_workshop.id, user_id: teacher.id)
+  end
+
+  enrollment = Pd::Enrollment.new(
+    pd_workshop_id: pd_workshop.id,
+    user_id: teacher.id,
+    first_name: teacher.name,
+    email: teacher.email,
+    created_at: created_at,
+  )
+
+  # apply a hack to bypass the last name and school info requirements; we only
+  # require those if the enrollment was created after a certain date, but we
+  # also examine whether or not the enrollment record has been persisted when
+  # considering that. Becuase these records were "created" before that date,
+  # they should pass validation but only _after_ they have been persisted in
+  # the db.
+  enrollment.save(validate: false)
+  enrollment.validate!
+
+  return enrollment
+end
+
+Workshop.where(pd_workshop_id: nil).each do |old_workshop|
+  next unless old_workshop.facilitators.count > 0
+
+  new_workshop = Pd::Workshop.new
+
+  # old workshops didn't have organizers, so pick one of the facilitators to be
+  # the new organizer
+  new_workshop.organizer = old_workshop.facilitators.first
+
+  new_workshop.location_name = old_workshop.location
+  new_workshop.created_at = old_workshop.created_at
+
+  # These values don't really matter anymore, but stick em in the notes field
+  # just for posterity
+  new_workshop.notes = "#{old_workshop.name}\n\n#{old_workshop.instructions}"
+
+  # Set Course and Subject based on ProgramType and Phase, which were the old
+  # equivalents.
+  new_workshop.course = PROGRAM_TYPES_TO_COURSES[old_workshop.program_type]
+  if subjects = PHASE_TO_SUBJECT[new_workshop.course]
+    key = old_workshop.phase
+    key = nil unless subjects.key?(key)
+    new_workshop.subject = subjects[key]
+  end
+
+  # capacity doesn't actually matter; set it to a number that will make sure
+  # all the enrollments will fit, and make sure it's not 0
+  new_workshop.capacity = [1, (old_workshop.attendances.count + old_workshop.teachers.count)].max
+
+  # all old workshops were paid but not on the map
+  new_workshop.funded = true
+  new_workshop.on_map = false
+
+  # funded CSF workshops were facilitator-funded
+  if new_workshop.course == Pd::SharedWorkshopConstants::COURSE_CSF
+    new_workshop.funding_type = 'facilitator'
+  end
+
+  new_workshop.save!
+
+  # facilitators => facilitators
+  old_workshop.facilitators.each do |facilitator|
+    new_workshop.facilitators << facilitator
+  end
+
+  # teachers => enrollments
+  old_workshop.teachers.each do |teacher|
+    get_or_create_enrollment(new_workshop, teacher, old_workshop.created_at)
+  end
+
+  # segments => sessions
+  old_workshop.segments.each do |segment|
+    session = Pd::Session.create!(
```

Comment:
the other places you create new objects, you checked first whether one already existed with the same data. should we do that here as well? (if this script runs more than once, are we risking creating duplicate sessions?)

Reply:
I don't think that's necessary right here; the most significant feature of a `Pd::Session` is its `pd_workshop_id`, since each Session is uniquely owned by a Workshop. In this case, the workshop has just been created, and we are careful to only create one new workshop for each old workshop (by relying on `old_workshop.pd_workshop_id = new_workshop.id`)